**Emmanuel Decena - ME EE**
Deep Learning Assignment

Build a classifier on CIFAR10 dataset using
- 3-layer MLP
- 3-layer +1 CNN

Compare performance

**1. Using 3-Layer MLP**

In [1]:
import tensorflow as tf
import numpy as np

from keras.datasets import cifar10
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers.core import Dense, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.callbacks import ModelCheckpoint
from keras.preprocessing.image import ImageDataGenerator
from keras.utils import to_categorical, plot_model

In [2]:
# load dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
num_labels = len(np.unique(y_train))
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

image_size = x_train.shape[1]
input_size = image_size * image_size * 3

# resize and normalize
x_train = np.reshape(x_train, [-1, input_size]) 
x_train = x_train.astype('float32') / 255
x_test = np.reshape(x_test, [-1, input_size])
x_test = x_test.astype('float32') / 255

In [3]:
# hyper-parameters
batch_size = 256
n_units = 256
epochs = 30

In [4]:
# model is a 3-layer MLP
model = Sequential(name='3-layer-MLP')
model.add(Dense(n_units, input_dim=input_size, activation='relu', name='input_layer'))
model.add(Dense(n_units, activation='relu', name='hidden_layer'))
model.add(Dense(num_labels, activation='softmax', name='output_layer'))
model.summary()

Model: "3-layer-MLP"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (Dense)          (None, 256)               786688    
_________________________________________________________________
hidden_layer (Dense)         (None, 256)               65792     
_________________________________________________________________
output_layer (Dense)         (None, 10)                2570      
Total params: 855,050
Trainable params: 855,050
Non-trainable params: 0
_________________________________________________________________


In [5]:
model.compile(loss="categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])

history = model.fit(
    x_train,
    y_train,
    validation_data=(x_test, y_test),
    epochs=epochs,
    batch_size=batch_size,
)

loss, acc = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=False)

print("\nTest accuracy: %.1f%%" % (100.0 * acc))


Epoch 1/30
196/196 [==============================] - 4s 19ms/step - loss: 2.0753 - accuracy: 0.2612 - val_loss: 1.9566 - val_accuracy: 0.2949
Epoch 2/30
196/196 [==============================] - 4s 19ms/step - loss: 1.8960 - accuracy: 0.3383 - val_loss: 1.8612 - val_accuracy: 0.3522
Epoch 3/30
196/196 [==============================] - 4s 20ms/step - loss: 1.8234 - accuracy: 0.3622 - val_loss: 1.8129 - val_accuracy: 0.3650
Epoch 4/30
196/196 [==============================] - 4s 20ms/step - loss: 1.7792 - accuracy: 0.3805 - val_loss: 1.7718 - val_accuracy: 0.3813
Epoch 5/30
196/196 [==============================] - 4s 21ms/step - loss: 1.7449 - accuracy: 0.3895 - val_loss: 1.7375 - val_accuracy: 0.3868
Epoch 6/30
196/196 [==============================] - 4s 21ms/step - loss: 1.7151 - accuracy: 0.4022 - val_loss: 1.7271 - val_accuracy: 0.4017
Epoch 7/30
196/196 [==============================] - 4s 21ms/step - loss: 1.6901 - accuracy: 0.4107 - val_loss: 1.6944 - val_accuracy: 0.4053

**2. Using 3-Layer CNN**

In [6]:
# load mnist dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

In [7]:
# input image dimensions
image_size = x_train.shape[1]
# resize and normalize
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

# network parameters
input_shape = (image_size, image_size, 3)
batch_size = 256
kernel_size = 3
pool_size = 3
filters = 32

In [8]:
model = Sequential()
model.add(Conv2D(filters=filters,
                 kernel_size=kernel_size, 
                 activation='relu', 
                 kernel_initializer='he_uniform', 
                 padding='same', 
                 input_shape=input_shape,
                 name='input_layer'))

model.add(MaxPooling2D(pool_size,name='hidden_layer'))
model.add(Flatten(name='hidden_layer2'))
model.add(Dense(10, activation='softmax',name='output_layer'))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (Conv2D)         (None, 32, 32, 32)        896       
_________________________________________________________________
hidden_layer (MaxPooling2D)  (None, 10, 10, 32)        0         
_________________________________________________________________
hidden_layer2 (Flatten)      (None, 3200)              0         
_________________________________________________________________
output_layer (Dense)         (None, 10)                32010     
Total params: 32,906
Trainable params: 32,906
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.compile(loss="categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])
# train the network
model.fit(x_train, y_train, epochs=epochs, batch_size=batch_size)

loss, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print("\nTest accuracy: %.1f%%" % (100.0 * acc))

Epoch 1/30
196/196 [==============================] - 22s 114ms/step - loss: 2.1054 - accuracy: 0.2499
Epoch 2/30
196/196 [==============================] - 20s 101ms/step - loss: 1.9082 - accuracy: 0.3414
Epoch 3/30
196/196 [==============================] - 19s 99ms/step - loss: 1.8153 - accuracy: 0.3774
Epoch 4/30
196/196 [==============================] - 19s 99ms/step - loss: 1.7509 - accuracy: 0.3999
Epoch 5/30
196/196 [==============================] - 19s 98ms/step - loss: 1.7011 - accuracy: 0.4202
Epoch 6/30
196/196 [==============================] - 21s 106ms/step - loss: 1.6595 - accuracy: 0.4336
Epoch 7/30
196/196 [==============================] - 25s 128ms/step - loss: 1.6211 - accuracy: 0.4495
Epoch 8/30
196/196 [==============================] - 26s 134ms/step - loss: 1.5878 - accuracy: 0.4591
Epoch 9/30
196/196 [==============================] - 24s 124ms/step - loss: 1.5589 - accuracy: 0.4683
Epoch 10/30
196/196 [==============================] - 26s 131ms/step - loss

###### **3. Performance**

The MLP model was outperformed by the CNN model using the same epochs and layers. 
The MLP's accuracy of ~47% was lower compared to the CNN accuracy of ~60%.
The MLP also used approximately ten times as much parameters at 855,050 compared to the CNN at 82,346.
The CNN model was more parameter efficient and has higher accuracy and thus is more suitable to use for this dataset.